sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,292 @@
+
+            # Calculate all costs from each point to all others in the cluster
+            in_cluster_all_costs = np.sum(in_cluster_distances, axis=1)
+
+            min_cost_idx = np.argmin(in_cluster_all_costs)
+            min_cost = in_cluster_all_costs[min_cost_idx]
+            curr_cost = in_cluster_all_costs[
+                np.argmax(cluster_k_idxs == medoid_idxs[k])]
+
+            # Adopt a new medoid if its distance is smaller then the current
+            if min_cost < curr_cost:
+                medoid_idxs[k] = cluster_k_idxs[min_cost_idx]
+
+    def transform(self, X):
+        """Transforms X to cluster-distance space.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            Data to transform.
+
+        Returns
+        -------
+        X_new : {array-like, sparse matrix}, shape=(n_samples, n_clusters)
+            X transformed in the new space of distances to cluster centers.
+        """
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+        check_is_fitted(self, "cluster_centers_")
+
+        return pairwise_distances(X, Y=self.cluster_centers_,
+                                  metric=self.metric)
+
+    def predict(self, X):
+        """Predict the closest cluster for each sample in X
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            New data to predict.
+
+        Returns
+        -------
+        labels : array, shape = (n_samples,)
+            Index of the cluster each sample belongs to.
+        """
+        check_is_fitted(self, "cluster_centers_")
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+
+        # Return data points to clusters based on which cluster assignment
+        # yields the smallest distance
+        return pairwise_distances_argmin(X, Y=self.cluster_centers_,
+                                         metric=self.metric)
+
+    def _compute_inertia(self, distances):
+        """Compute inertia of new samples. Inertia is defined as the sum of the
+        sample distances to closest cluster centers.
+
+        Parameters
+        ----------
+        distances : {array-like, sparse matrix}, shape=(n_samples, n_clusters)
+            Distances to cluster centers.
+
+        Returns
+        -------
+        Sum of sample distances to closest cluster centers.
+        """
+
+        # Define inertia as the sum of the sample-distances
+        # to closest cluster centers
+        inertia = np.sum(np.min(distances, axis=1))
+
+        return inertia
+
+    def _initialize_medoids(self, D, n_clusters):
+        """Select initial mediods randomly or heuristically"""
+
+        if self.init == 'random':  # Random initialization
+            # Pick random k medoids as the initial ones.
+            medoids = self.random_state_.choice(len(D), n_clusters)
+        else:  # Initialization by heuristic
```

Comment:
Is this heuristic something from the literature? it seems to me very likely to pick medoids in the same cluster. Certainly, this seems to be the case when I try it on `make_blobs()`. Is this ever faster to converge than random??

@amueller knows something about kmeans initialisation, I suspect. Any thoughts on a better-than-random heuristic initialisation strategy?

A greedy strategy that might be better than this one (and could be better than random for imbalanced clusters): Choose one point (perhaps the most central, like here), then choose the point farthest from it, then choose the point farthest from the first two, etc??

Reply:
A comparison of random and heuristic in terms of wall time to convergence, and inertia at convergence, would be worth doing...

Reply:
I was thinking that it would be good to use something similar to KMeans++, which is the default initialization strategy in KMeans (It tries to select initial centers that are far apart).

Reply:
I don't think we should aim to innovate too much. If this is described in the literature, then okay. But I've also not looked into how kmeans++ works.

Reply:
Not only is it described in the literature, it's the default initialization method in KMeans: (copied from the k_means_._k_init "Notes" section.

    Selects initial cluster centers for k-mean clustering in a smart way
    to speed up convergence. see: Arthur, D. and Vassilvitskii, S.
    "k-means++: the advantages of careful seeding". ACM-SIAM symposium
    on Discrete algorithms. 2007

The code can mostly be copy-pasted over from KMeans, and I was thinking of doing the same thing with the docstring. I can think of at least one simple example that showcases the method's usefulness.

Reply:
Yes, I realise it's described in the KMeans literature. I'd not reviewed it to see if applicable to KMedoids. If we can adapt or reuse it, then great.